File: blog-frontend/src/types/post.ts
export interface Post {
  id: string;
  title: string;
  slug: string;
  category: string;
  author_name: string;
  published_at: string;
  created_at: string;

  likes: number;
  dislikes: number;
  comments: number;

  // Detail fields (optional because list API may not return them)
  content?: string;
  image?: string;
}

export interface Pagination {
  page: number;
  per_page: number;
  total: number;
  total_pages: number;
}

export interface PostsResponse {
  posts: Post[];
  pagination: Pagination;
}


// Detailed Post Types

export interface PostAuthor {
  id: string;
  username: string;
}

export interface PostReactions {
  likes: number;
  dislikes: number;
}

export interface CommentUser {
  id: string;
  username: string;
}

export interface PostComment {
  id: string;
  post_id: string;
  content: string;
  created_at: string;

  user: CommentUser;
  replies: PostComment[];
}

export type ReactionType = "like" | "dislike";


export interface PostDetail {
  id: string;
  title: string;
  slug: string;
  category: string;
  content: string;

  published_at: string | null;
  created_at: string;
  updated_at: string | null;

  author: PostAuthor;

  reactions: PostReactions;
  user_reaction: ReactionType | null;

  comment_count: number;
  comments: PostComment[];
}

// Create post
export interface CreatePostPayload {
  title: string;
  slug: string;
  content: string;
  category: string;
}

export interface CreatedPost {
  id: string;
  title: string;
  slug: string;
  category: string;
  status: string;
  created_at: string;
}

export interface CreatePostResponse {
  message: string;
  post: CreatedPost;
}

export interface ApiErrorResponse {
  message: string;
}

File: blog_backend/app/schemas/post.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.core.enums import ReactionType


class PostListItem(BaseModel):
    id: UUID
    title: str
    slug: str
    category: str
    author_name: str
    published_at: datetime | None
    created_at: datetime

    likes: int
    dislikes: int
    comments: int

    class Config:
        from_attributes = True


class PaginationMeta(BaseModel):
    page: int
    per_page: int
    total: int
    total_pages: int


class PostListResponse(BaseModel):
    posts: list[PostListItem]
    pagination: PaginationMeta


class UserSummary(BaseModel):
    id: UUID
    username: str | None

    class Config:
        from_attributes = True


class PostAuthor(UserSummary):
    pass


class PostReactions(BaseModel):
    likes: int
    dislikes: int


class CommentOut(BaseModel):
    id: UUID
    post_id: UUID
    content: str
    created_at: datetime

    user: UserSummary
    replies: list["CommentOut"] = Field(default_factory=list)

    class Config:
        from_attributes = True


CommentOut.model_rebuild()


class PostDetailResponse(BaseModel):
    id: UUID
    title: str
    slug: str
    category: str
    content: str

    published_at: datetime | None
    created_at: datetime
    updated_at: datetime | None

    author: PostAuthor
    reactions: PostReactions
    user_reaction: ReactionType | None = None

    comment_count: int
    comments: list[CommentOut]

    class Config:
        from_attributes = True


class PostCreate(BaseModel):
    title: str = Field(min_length=3, max_length=255)
    slug: str = Field(min_length=3, max_length=255)
    content: str
    category: str


class PendingPostResponse(BaseModel):
    id: UUID
    title: str
    slug: str
    category: str
    status: str
    created_at: datetime

    class Config:
        from_attributes = True


class CreatePostResponse(BaseModel):
    message: str
    post: PendingPostResponse


class DeletePostResponse(BaseModel):
    message: str
    deleted_post_id: UUID